spring-xd-dirt/src/main/java/org/springframework/xd/dirt/stream/XDStreamParser.java

```diff
@@ -322,4 +333,98 @@ private ModuleType resolveModuleType(String moduleName, ModuleType... candidates
 		throw new NoSuchModuleException(moduleName, candidates);
 	}
 
+	/**
+	 * A wrapper around the canonical parser that supports multiline "document" parsing.
+	 *
+	 * <p>While XDStreamParser parses a single definition, this class is able to parse multiple, reporting
+	 * errors or accepting constructs like taps against definitions that appear earlier in the document (without
+	 * actually saving them in the main repository).</p>
+	 *
+	 * @author Eric Bottard
+	 */
+	public static class MultiLineDocumentParser {
+
+		private final XDStreamParser delegate;
+
+		public MultiLineDocumentParser(XDStreamParser delegate) {
+			this.delegate = delegate;
+		}
+
+		public DocumentParseResult parse(String[] document) {
+			DocumentParseResult result = new DocumentParseResult(document.length);
+			CrudRepository<BaseDefinition, String> transientRepository = new TransientDefinitionRepository();
+			StreamConfigParser parser = new StreamConfigParser(transientRepository);
+			for (String nameAndDefinitionPair : document) {
+				try {
+
+					StreamNode stream = parser.parse(nameAndDefinitionPair);
+					String streamName = stream.getStreamName();
+					List<ModuleDescriptor> moduleDescriptors = delegate.buildModuleDescriptors(streamName,
+							nameAndDefinitionPair, ParsingContext.stream, stream);
+					BaseDefinition streamDefinition = new StreamDefinition(streamName, nameAndDefinitionPair);
+					transientRepository.save(streamDefinition);
+					result.success(moduleDescriptors);
+				}
+				catch (Exception e) {
+					result.failure(e);
+				}
+			}
+			return result;
+		}
+
+
+		/**
+		 * A throwaway repository that is used during multi-definition parsing. "Sees" definitions that are
+		 * known to the real {@link #repository} and stores new definitions in an in-memory map.
+		 *
+		 * @author Eric Bottard
+		 */
+		private class TransientDefinitionRepository
+				extends AbstractRepository<BaseDefinition, String>
+				implements CrudRepository<BaseDefinition, String> {
+
+			private Map<String, BaseDefinition> store = new HashMap<>();
+
+			@Override
+			public <S extends BaseDefinition> S save(S entity) {
+				store.put(entity.getName(), entity);
+				return entity;
+			}
+
+			@Override
+			public BaseDefinition findOne(String s) {
+				BaseDefinition inMemory = store.get(s);
+				return inMemory != null ? inMemory : delegate.repository.findOne(s);
+			}
+
+			@Override
+			public Iterable<BaseDefinition> findAll() {
+				return Iterables.concat(delegate.repository.findAll(), store.values());
+			}
+
+			@Override
+			public long count() {
+				return store.size() + delegate.repository.count();
```

Comment:
This could contain duplicate count as the stream definitions being parsed are always stored in this store. In any case, I don't see any of these methods (except `save` and `findOne`) being used. It doesn't hurt.